File: Components/Strata.Client.Core/main/java/strata/client/core/service/AbstractRestClient.java
//////////////////////////////////////////////////////////////////////////////
// AbstractRestClient.java
//////////////////////////////////////////////////////////////////////////////

package strata.client.core.service;

import jakarta.ws.rs.client.*;
import jakarta.ws.rs.core.MediaType;
import jakarta.ws.rs.core.MultivaluedHashMap;
import jakarta.ws.rs.core.MultivaluedMap;
import jakarta.ws.rs.core.Response;

import java.util.Map;
import java.util.concurrent.CompletionStage;

public abstract
class AbstractRestClient
{
    private Client                        itsClient;
    private WebTarget                     itsBaseTarget;
    private MultivaluedMap<String,Object> itsHeaders;
    private IResponseProcessor            itsResponseProcessor;

    protected
    AbstractRestClient(
        ClientBuilder builder,
        String        baseUrl,
        String        endpointPath)
    {
        this(builder,baseUrl,endpointPath,new StandardResponseProcessor());
    }

    protected
    AbstractRestClient(
        Client client,
        String baseUrl,
        String endpointPath)
    {
        this(client,baseUrl,endpointPath,new StandardResponseProcessor());
    }

    protected
    AbstractRestClient(
        ClientBuilder      builder,
        String             baseUrl,
        String             endpointPath,
        IResponseProcessor processor)
    {
        itsClient =
                builder
                    .register(new ObjectMapperProvider())
                    //.register(new ObjectMapperContextResolver())
                    .build();

        itsBaseTarget = itsClient.target(initialize(baseUrl,endpointPath));
        itsHeaders = new MultivaluedHashMap<>();
        itsResponseProcessor = processor;
    }

    protected
    AbstractRestClient(
        Client             client,
        String             baseUrl,
        String             endpointPath,
        IResponseProcessor processor)
    {
        itsClient = client;
        itsBaseTarget = itsClient.target(initialize(baseUrl,endpointPath));
        itsHeaders = new MultivaluedHashMap<>();
        itsResponseProcessor = processor;
    }

    public AbstractRestClient
    setHeader(String headerKey,String headerValue)
    {
        itsHeaders.add(headerKey,headerValue);
        return this;
    }

    public AbstractRestClient
    clearHeader(String headerKey)
    {
        itsHeaders.remove(headerKey);
        return this;
    }

    public AbstractRestClient
    clearHeaders()
    {
        itsHeaders.clear();
        return this;
    }

    public boolean
    hasHeader(String headerKey)
    {
        return itsHeaders.containsKey(headerKey);
    }

    public void
    close()
    {
        itsClient.close();
    }

    protected <Request,Reply> Reply
    doPost(String methodPath,Class<Reply> replyType,Request request)
    {
        return
            itsResponseProcessor.process(
                replyType,
                toResponse(
                    methodPath,
                    buildRequest(methodPath).post(Entity.json(request))));
    }

    protected <Request,Reply> Reply
    doPut(String methodPath,Class<Reply> replyType,Request request)
    {
        return
            itsResponseProcessor.process(
                replyType,
                toResponse(
                    methodPath,
                    buildRequest(methodPath)
                        .put(
                            Entity.entity(
                                request,
                                MediaType.APPLICATION_JSON))));
    }

    protected <Request,Reply> Reply
    doDelete(String methodPath,Class<Reply> replyType,Map<String,Object> params)
    {
        WebTarget target = itsBaseTarget.path(methodPath);

        for (Map.Entry<String,Object> param:params.entrySet())
            target.queryParam(param.getKey(),param.getValue());

        return
            itsResponseProcessor.process(
                replyType,
                toResponse(
                    methodPath,
                    target
                        .request(MediaType.APPLICATION_JSON)
                        .buildDelete()
                        .invoke()));
    }

    protected <Request,Reply> Reply
    doGet(String methodPath,Class<Reply> replyType,Map<String,Object> params)
    {
        WebTarget target = itsBaseTarget.path(methodPath);

        for (Map.Entry<String,Object> param:params.entrySet())
            target.queryParam(param.getKey(),param.getValue());

        return
            itsResponseProcessor.process(
                replyType,
                toResponse(
                    methodPath,
                    target
                        .request(MediaType.APPLICATION_JSON)
                        .buildGet()
                        .invoke()));
    }

    protected <Request,Reply> CompletionStage<Reply>
    doPostAsync(String methodPath,Class<Reply> replyType,Request request)
    {
        return
            buildRequest(methodPath)
                .rx()
                .post(Entity.json(request))
                .thenApply(
                    response ->
                        itsResponseProcessor.process(
                            replyType,
                            toResponse(methodPath,response)));
    }

    protected <Request,Reply> CompletionStage<Reply>
    doPutAsync(String methodPath,Class<Reply> replyType,Request request)
    {
        return
            buildRequest(methodPath)
                .rx()
                .put(
                    Entity.entity(
                        request,
                        MediaType.APPLICATION_JSON))
                .thenApply(
                    response ->
                        itsResponseProcessor.process(
                            replyType,
                            toResponse(methodPath,response)));
    }

    protected <Request,Reply> CompletionStage<Reply>
    doDeleteAsync(
        String             methodPath,
        Class<Reply>       replyType,
        Map<String,Object> pathParams,
        Map<String,Object> queryParams)
    {
        WebTarget target =
            itsBaseTarget
                .path(methodPath)
                .resolveTemplates(pathParams);

        for (Map.Entry<String,Object> param:queryParams.entrySet())
            target.queryParam(param.getKey(),param.getValue());

        return
            target
                .request(MediaType.APPLICATION_JSON)
                .rx()
                .delete()
                .thenApply(
                    response ->
                        itsResponseProcessor.process(
                            replyType,
                            toResponse(methodPath,response)));
    }

    protected <Request,Reply> CompletionStage<Reply>
    doGetAsync(
        String             methodPath,
        Class<Reply>       replyType,
        Map<String,Object> pathParams,
        Map<String,Object> queryParams)
    {
        WebTarget target =
            itsBaseTarget
                .path(methodPath)
                .resolveTemplates(pathParams);

        for (Map.Entry<String,Object> param:queryParams.entrySet())
            target.queryParam(param.getKey(),param.getValue());

        return
            target
                .request(MediaType.APPLICATION_JSON)
                .rx()
                .get()
                .thenApply(
                    response ->
                        itsResponseProcessor.process(
                            replyType,
                            toResponse(methodPath,response)));
    }

    private Invocation.Builder
    buildRequest(String path)
    {
        return
            itsBaseTarget
                .path(path)
                .request(MediaType.APPLICATION_JSON)
                .headers(itsHeaders)
                .accept(MediaType.APPLICATION_JSON);
    }

    private static String
    initialize(String baseUrl,String endpointPath)
    {
        return
            baseUrl.endsWith("/" + endpointPath)
                ? baseUrl
                : baseUrl + "/" + endpointPath;
    }

    private IResponse
    toResponse(String methodPath,Response response)
    {
        return
            new StandardResponse(
                itsBaseTarget.getUri().toString(),
                methodPath,
                response);
    }
}

//////////////////////////////////////////////////////////////////////////////

File: Components/Strata.Client.Core/main/typescript/Service/AbstractRestClient.ts
import {CompletableObservable,ICompletionStage} from "strata.foundation.core/Concurrent";
import {IResponseProcessor} from "./IResponseProcessor";
import {StandardResponseProcessor} from "./StandardResponseProcessor";
import {StandardResponse} from "./StandardResponse";
import {AbstractServiceReply} from "strata.foundation.core/Transfer";
import {ExceptionDataBuilder} from "strata.foundation.core/Transfer";
import {IHeadersConsumer} from "./IHeadersConsumer";

export
abstract class AbstractRestClient
{
    private readonly baseUrl: string;
    private headers: Headers;
    private consumer: IHeadersConsumer;
    private readonly responseProcessor: IResponseProcessor;

    protected constructor(baseUrl: string,responseProcessor?: IResponseProcessor)
    {
        this.baseUrl = baseUrl;
        this.headers = new Headers();
        this.consumer = null;
        this.responseProcessor =
            responseProcessor == null
                ? new StandardResponseProcessor()
                : responseProcessor;
    }

    protected setHeader(headerKey: string,headerValue: string): AbstractRestClient
    {
        this.headers.append(headerKey,headerValue);
        return this;
    }

    protected clearHeader(headerKey: string): AbstractRestClient
    {
        this.headers.delete(headerKey);
        return this;
    }

    protected clearHeaders(): AbstractRestClient
    {
        this.headers = new Headers();
        return this;
    }

    protected getHeader(headerKey: string): string
    {
        return this.headers.get(headerKey);
    }

    protected hasHeader(headerKey: string): boolean
    {
        return this.headers.has(headerKey);
    }

    protected setHeadersConsumer(consumer: IHeadersConsumer): void
    {
        this.consumer = consumer;
    }

    protected doPost<TReply,TRequest>(methodPath: string,request: TRequest):
        ICompletionStage<TReply>
    {
        return this.sendRequestWithBody("POST",methodPath,request);
    }

    protected doPut<TReply,TRequest>(methodPath: string,request: TRequest):
        ICompletionStage<TReply>
    {
        return this.sendRequestWithBody("PUT",methodPath,request);
    }

    protected doDelete<TReply>(
        methodPath: string,
        pathParams: Map<string,Object>,
        queryParams: Map<string,Object>):
        ICompletionStage<TReply>
    {
        return this.sendRequestWithParameters("DELETE",methodPath,pathParams,queryParams);
    }

    protected doGet<TReply>(
        methodPath: string,
        pathParams: Map<string,Object>,
        queryParams: Map<string,Object>):
        ICompletionStage<TReply>
    {
        return this.sendRequestWithParameters("GET",methodPath,pathParams,queryParams);
    }

    private sendRequestWithBody<TReply,TRequest>(
        method: string,
        methodPath: string,
        request: TRequest):
        ICompletionStage<TReply>
    {
        return CompletableObservable.fromPromise(
            fetch(
                this.getUrl(methodPath),
                {
                    method: method,
                    mode: "cors",
                    body: JSON.stringify(request),
                    headers: this.headers})
                .then(response => this.processHeaders(response))
                .then(response => this.processResponse<TReply>(response,methodPath))
                .catch(exception => this.processException<TReply>(exception)));
    }

    private sendRequestWithParameters<TReply>(
        method: string,
        methodPath: string,
        pathParams: Map<string,Object>,
        queryParams: Map<string,Object>):
        ICompletionStage<TReply>
    {
        let requestUrl: string =
            this.resolveTemplates(this.getUrl(methodPath),pathParams);

        requestUrl = this.appendQueryParams(requestUrl,queryParams);

        return CompletableObservable.fromPromise(
            fetch(
                requestUrl,
                {
                    method: method,
                    mode: "cors",
                    headers: this.headers})
                .then(response => this.processHeaders(response))
                .then(response => this.processResponse<TReply>(response,methodPath))
                .catch(exception => this.processException<TReply>(exception)));
    }

    private getUrl(methodPath: string): string
    {
        return this.baseUrl + "/" + methodPath;
    }

    private resolveTemplates(inputUrl: string,pathParams: Map<string,Object>): string
    {
        pathParams
            .forEach(
                (value,key,ignore) =>
                {
                    let template: string = '{' + key + '}';
                    return inputUrl = inputUrl.replace(template,value.toString);
                });

        return inputUrl;
    }

    private appendQueryParams(inputUrl: string,queryParams: Map<string,Object>): string
    {
        let i: number = 0;

        queryParams
            .forEach(
                (value,key,ignore) =>
                {
                    if (i == 0)
                        inputUrl = inputUrl + '?';
                    else
                        inputUrl = inputUrl + '&';

                    ++i;
                    return inputUrl = inputUrl + key + '=' + value.toString();
                });

        return inputUrl;
    }

    private processHeaders(response: Response): Response
    {
        if (this.consumer)
            this.consumer.accept(response.headers);

        return response;
    }

    private processResponse<TReply>(response: Response,methodPath: string):
        Promise<TReply>
    {

        console.log("AbstractRestClient.processResponse")
        return this
            .responseProcessor
            .process<TReply>(new StandardResponse(this.baseUrl,methodPath,response));
    }

    private processException<TReply>(exception: any): TReply
    {
        const reply: TReply =  {} as TReply;

        console.log("AbstractRestClient.processException: " + exception);

        if (this.isAbstractServiceReply(reply))
        {
            const serviceReply: AbstractServiceReply = reply as AbstractServiceReply;

            console.log("Service reply is subtype of AbstractServiceReply");
            serviceReply.exception =
                    new ExceptionDataBuilder()
                        .setExceptionType("strata.client.core.service.ServiceException")
                        .setExceptionMessage(exception.message)
                        .build();
            serviceReply.success = false;
            serviceReply.failureMessage = exception.message;
        }

        return reply;
    }

    private isAbstractServiceReply<TReply>(reply: TReply): boolean
    {
        const comparable: AbstractServiceReply = {} as AbstractServiceReply;

        return Object
            .keys(comparable)
            .every((property) => reply.hasOwnProperty(property));
    }
}